### PR TITLE
Migrate emails to inertia

### PR DESCRIPTION
**Pull request by stefanbinoj**
Part of #856 
Prev PR #2417 

Migrate following routes to inertia : 
- Published emails page (/emails/published)
- Scheduled emails page (/emails/scheduled)
- Draft emails page (/emails/draft)

## Changes
- Implemented infinite scroll using inertia (updated gem version)
- Better DRY-ed up state with common imports
- Also optimizes db calls by addressing some `Bullet Warnings` 
- Removed corressponding internal api and ts helpers

## Before 

https://github.com/user-attachments/assets/45a080c0-ebc5-452d-a743-eec70ffb6fbb

https://github.com/user-attachments/assets/9bc5cfcd-c156-44e2-b933-e024ac8d04d7

### Infinite Scroll Working fine with query : 

https://github.com/user-attachments/assets/8e1b40e2-0056-4875-bac6-0ce8145ebad4

## AI disclosure
Used Claude Opus 4.5 via Opencode for initial drafting of this PR

## Confirmation
I have watched all Antiwork Livestreams

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -110,7 +110,7 @@ gem "httparty", "~> 0.21"
 gem "ibandit", "~> 1.26"
```
Comment (stefanbinoj) [inline Gemfile:110]: manually updated the version here and then ran : 
```bash
bundle install 
```

File: app/javascript/components/EmailsPage/Layout.tsx [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+            <Link href={Routes.published_emails_path()}>Published</Link>
+          </Tab>
+          <Tab asChild isSelected={selectedTab === "scheduled"}>
+            <Link href={Routes.scheduled_emails_path()}>Scheduled</Link>
+          </Tab>
+          <Tab asChild isSelected={selectedTab === "drafts"}>
+            <Link href={Routes.drafts_emails_path()}>Drafts</Link>
+          </Tab>
+          {/* Legacy pages - use regular href for full page reload */}
+          <Tab href={Routes.followers_path()} isSelected={selectedTab === "subscribers"}>
```
Comment (stefanbinoj) [inline app/javascript/components/EmailsPage/Layout.tsx:92]: Not migrated to inertia still uses `href`

File: app/javascript/hooks/useDebouncedSearch.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+ * Hook for managing search query state with debounced Inertia reload.
+ *
+ * - Maintains local query state for immediate UI updates
+ * - Debounces the actual server request by 500ms
+ * - Only triggers reload when query actually changes (not on mount)
+ * - Resets to page 1 when search query changes
+ * - Resets installments prop to clear merged infinite scroll data
+ */
+export function useDebouncedSearch() {
+  const [query, setQuery] = React.useState(() => new URLSearchParams(window.location.search).get("query") || "");
```
Comment (stefanbinoj) [inline app/javascript/hooks/useDebouncedSearch.ts:17]: Letting state to get hydrated from the server query, when user reloads

File: app/javascript/pages/Emails/Drafts.tsx [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+  const [selectedInstallment, setSelectedInstallment] = React.useState<DraftInstallment | null>(null);
+  const [installmentToDelete, setInstallmentToDelete] = React.useState<DraftInstallment | null>(null);
+
+  return (
+    <EmailsLayout selectedTab="drafts" hasPosts={!!installments.length} query={query} onQueryChange={setQuery}>
+      <div className="space-y-4 p-4 md:p-8">
+        {installments.length > 0 ? (
+          <>
+            <InfiniteScroll data="installments" preserveUrl>
```
Comment (stefanbinoj) [inline app/javascript/pages/Emails/Drafts.tsx:73]: preserverUrl for displaying `?query=`

File: app/javascript/components/EmailsPage/DeleteEmailModal.tsx [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  onClose,
+  onSuccess,
+  confirmationMessage,
+}: Props<T>) {
+  const deleteForm = useForm({});
+
+  const handleDelete = () => {
+    if (!installment) return;
+    deleteForm.delete(Routes.email_path(installment.external_id), {
+      only: ["flash"],
```
Comment (stefanbinoj) [inline app/javascript/components/EmailsPage/DeleteEmailModal.tsx:37]: we only listens for flash message

File: app/javascript/pages/Emails/Drafts.tsx [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+                </div>
+              </Sheet>
+            ) : null}
+            <DeleteEmailModal
+              installment={installmentToDelete}
+              onClose={() => setInstallmentToDelete(null)}
+              onSuccess={(deleted) => {
+                setSelectedInstallment(null);
+                router.replaceProp("installments", (current: DraftInstallment[]) =>
+                  current.filter((i) => i.external_id !== deleted.external_id),
```
Comment (stefanbinoj) [inline app/javascript/pages/Emails/Drafts.tsx:154]: Since server only responds with `notice`. This `onSuccess` handler replaces the currently fetched installments
```tsx
router.replaceProp("installments", (current: DraftInstallment[]) =>
                  current.filter((i) => i.external_id !== deleted.external_id),
                  
                  ```

File: app/controllers/emails_controller.rb [hunk 7]
```diff
@@ -1,21 +1,65 @@
...
+  end
+
+  def published
+    authorize Installment, :index?
     create_user_event("emails_view")
 
-    if request.path == emails_path
-      default_tab = Installment.alive.not_workflow_installment.scheduled.where(seller: current_seller).exists? ? "scheduled" : "published"
-      redirect_to "#{emails_path}/#{default_tab}", status: :moved_permanently
+    @title = "Published Emails"
```
Comment (stefanbinoj) [inline app/controllers/emails_controller.rb:26]: Better title since we have more control now

File: app/controllers/emails_controller.rb [hunk 8]
```diff
@@ -1,21 +1,65 @@
...
+    presenter = PaginatedInstallmentsPresenter.new(seller: current_seller, type: Installment::PUBLISHED, page: params[:page], query: params[:query])
```
Comment (stefanbinoj) [inline app/controllers/emails_controller.rb:27]: We now pass page and query from request as well. But the page is handled internally by inertia infiniteScroll

File: app/controllers/emails_controller.rb [hunk 9]
```diff
@@ -1,21 +1,65 @@
...
+  end
+
+  def destroy
+    authorize @installment
+
+    if @installment.mark_deleted
+      @installment.installment_rule&.mark_deleted!
+      redirect_back fallback_location: published_emails_path, notice: "Email deleted!", status: :see_other
+    else
+      redirect_back fallback_location: published_emails_path, alert: "Sorry, something went wrong. Please try again.", status: :see_other
```
Comment (stefanbinoj) [inline app/controllers/emails_controller.rb:56]: Now we have a new endpoint for deleting published and scheduled emails! Removed dependency on internal api. 
Also since both pages (/published and /scheduled) uses same destroy endpoint added `redirect_back`
```ruby
redirect_back fallback_location: published_emails_path ... 
```

File: app/presenters/paginated_installments_presenter.rb [hunk 10]
```diff
@@ -17,11 +17,13 @@ def initialize(seller:, type:, page: nil, query: nil)
 
   def props
     if query.blank?
-      installments = Installment.includes(:installment_rule).all
+      # Always include :seller (needed for full_url -> user -> seller.presence)
+      # Only include :installment_rule for scheduled/draft (not needed for published)
+      installments = Installment.includes(:seller)
+      installments = installments.includes(:installment_rule) if type != Installment::PUBLISHED
```
Comment (stefanbinoj) [inline app/presenters/paginated_installments_presenter.rb:23]: Previously we were querying for `installment_rule` for both /published and /scheduled but /pubshied doesn't need that 



Bullet Warning
user: stefanbinoj AVOID eager loading detected Installment => [:installment_rule] Remove from your query: .includes([:installment_rule])

File: app/presenters/paginated_installments_presenter.rb [hunk 11]
```diff
@@ -17,11 +17,13 @@ def initialize(seller:, type:, page: nil, query: nil)
...
       installments = installments.ordered_updates(seller, type).public_send(type)
       installments = installments.unscope(:order).order("installment_rules.to_be_published_at ASC") if type == Installment::SCHEDULED
-      pagination, installments = pagy(installments, page:, limit: PER_PAGE, overflow: :empty_page)
-      pagiation_metadata = { count: pagination.count, next: pagination.next }
+      pagination, installments = pagy_countless(installments, page:, limit: PER_PAGE, overflow: :empty_page)
```
Comment (stefanbinoj) [inline app/presenters/paginated_installments_presenter.rb:26]: using `pagy_counless` for inertia infinite scroll

File: app/presenters/paginated_installments_presenter.rb [hunk 12]
```diff
@@ -36,14 +38,23 @@ def props
...
+        next_page: can_paginate_further ? page + 1 : nil
+      }
     end
 
+    installments_data = installments.map { InstallmentPresenter.new(seller:, installment: _1).props }
+
     {
-      installments: installments.map { InstallmentPresenter.new(seller:, installment: _1).props },
-      pagination: pagiation_metadata,
+      installments: InertiaRails.scroll(pagination) { installments_data },
```
Comment (stefanbinoj) [inline app/presenters/paginated_installments_presenter.rb:57]: We manually dont have to manage `pagination_metadata` `InertiaRails.scroll(pagination)` will do it now

File: app/presenters/paginated_installments_presenter.rb [hunk 13]
```diff
@@ -17,11 +17,13 @@ def initialize(seller:, type:, page: nil, query: nil)
 
   def props
     if query.blank?
-      installments = Installment.includes(:installment_rule).all
+      # Always include :seller (needed for full_url -> user -> seller.presence)
+      # Only include :installment_rule for scheduled/draft (not needed for published)
+      installments = Installment.includes(:seller)
```
Comment (stefanbinoj) [inline app/presenters/paginated_installments_presenter.rb:22]: Previously In `/published` we,  after loading installments, we're accessing installment.seller for each installment without having eager loaded the seller association. This causes N+1 queries.

Bullet Warning
user: stefanbinoj USE eager loading detected Installment => [:seller] Add to your query: .includes([:seller])

File: config/routes.rb [hunk 14]
```diff
@@ -802,7 +801,16 @@ def product_info_and_purchase_routes(named_routes: true)
...
+    # TODO: Add :new, :edit routes once those pages are migrated to Inertia
+    resources :emails, only: [:index, :destroy] do
+      collection do
+        get :published
+        get :scheduled
+        get :drafts
+      end
+    end
+    # TODO: Remove this catch-all once new, edit email pages are migrated to Inertia
+    get "/emails/*other", to: "emails#index" # route handled by react-router
```
Comment (stefanbinoj) [inline config/routes.rb:813]: Now since we have dedicated /published and /scheduled routes it is better to decrement the priority of this line (or else we will get in a redirection loop)

**Comment by stefanbinoj** [comment]
Some tests will be failing working on it!

**Comment by Devwill2** [comment]
Hi @EmCousin,
I wanted to share a concern about the recent PR approval. I’ve been working on this PR daily for several weeks and implemented the changes based on the review feedback you suggested, including splitting the PR.
However, I noticed that @stefanbinoj opened a smaller PR with similar changes (based on the same feedback), and that PR was approved. This feels a bit discouraging from my side, since a significant amount of time and effort went into my work, but it seems like the credit may go elsewhere simply because the PR is smaller.
I completely understand the preference for smaller PRs, but in this case, the work and ideas were already in progress and reviewed. I just wanted to highlight how this impacts time investment and contributions, and ask how you’d suggest handling situations like this in the future so efforts don’t get duplicated or overlooked.
Thanks for hearing me out.

**Comment by 4arjun** [comment]
> However, I noticed that @stefanbinoj opened a smaller PR with similar changes (based on the same feedback), and that PR was approved. This feels a bit discouraging from my side, since a significant amount of time and effort went into my work, but it seems like the credit may go elsewhere simply because the PR is smaller.


Thank you for raising this concern, Devwill, I can relate to it.

I had already been working on the Inertia migration for Library, Reviews, Wishlists, and Following Wishlists for over a month. The Reviews page migration was part of that work.
(See https://github.com/antiwork/gumroad/pull/2287)

Later, the Reviews scope was split out and merged separately while that work was already in progress (by @Devwill2). This situation is quite similar to what is being discussed here.

I am not opposed to resourcing the migration if a smaller or cleaner implementation leads to better code being merged. That would be a positive outcome.

**Comment by Devwill2** [comment]
> > However, I noticed that @stefanbinoj opened a smaller PR with similar changes (based on the same feedback), and that PR was approved. This feels a bit discouraging from my side, since a significant amount of time and effort went into my work, but it seems like the credit may go elsewhere simply because the PR is smaller.
> 
> Thank you for raising this concern, Devwill, I can relate to it.
> 
> I had already been working on the Inertia migration for Library, Reviews, Wishlists, and Following Wishlists for over a month. The Reviews page migration was part of that work. (See #2287)
> 
> Later, the Reviews scope was split out and merged separately while that work was already in progress (by @Devwill2). This situation is quite similar to what is being discussed here.
> 
> I am not opposed to resourcing the migration if a smaller or cleaner implementation leads to better code being merged. That would be a positive outcome.

@4arjun Thanks for sharing the context — I understand your point.
Just to clarify from my side: I genuinely wasn’t aware that you were already working on the Reviews JS migration at that time. If I had known, or if it had been communicated on my PR, I would have happily closed my PR. I’ve done that in the past as well when overlaps were identified.
My broader concern (and what I was trying to raise with @EmCousin and @ershad ) is more general and about future situations. I’ve been working daily on this PR for weeks, responding to reviews and making changes as suggested. Then someone else creates a new PR by following the same review feedback from my PR, and that PR gets approved and merged. In cases like this, it feels discouraging because a lot of time and effort has already been invested, but the credit and outcome shift simply because the new PR is smaller.
I fully agree that better and cleaner code should always be merged — that’s not the issue. The concern is about duplicated effort and how priority is handled when someone is already actively working on a scope.
If there were clearer rules or guidance around ownership, overlap, or how to handle such situations (for example, communicating early or giving priority to ongoing work), it would help avoid this kind of frustration going forward.
Thanks for hearing me out, and I appreciate the discussion.

File: app/controllers/api/internal/installments_controller.rb [hunk 15]
```diff
@@ -30,12 +30,21 @@ def update
     def save_installment
       service = SaveInstallmentService.new(seller: current_seller, params:, installment: @installment, preview_email_recipient: impersonating_user || logged_in_user)
       if service.process
+        set_flash_for_installment_save(service.installment)
```
Comment (stefanbinoj) [inline app/controllers/api/internal/installments_controller.rb:33]: This is necessary because currently on client side we redirect user to /published or /scheduled upon successfull email action using `window.location.href`. Thus setting alert on client side using `showAlert` just before redirection will not show alert as page is fully reloaded. 

This fix of updating respective rails controller for saving installment to have `flash[:notice] = ` enables inertia/layout to listen for flash messages using `usePageHooks` and show the flash even while page reloading

File: app/javascript/components/server-components/EmailsPage/EmailForm.tsx [hunk 16]
```diff
@@ -586,29 +586,27 @@ export const EmailForm = () => {
...
-            ? "Preview link opened."
-            : action === "save_and_schedule"
-              ? "Email successfully scheduled!"
-              : action === "save_and_publish"
-                ? `Email successfully ${channel.profile ? "published" : "sent"}!`
-                : installment?.external_id
-                  ? "Changes saved!"
-                  : "Email created!",
-        "success",
-      );
```
Comment (stefanbinoj) [inline app/javascript/components/server-components/EmailsPage/EmailForm.tsx:602]: https://github.com/antiwork/gumroad/pull/2436/files#r2643713640

File: app/javascript/components/EmailsPage/Layout.tsx [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+    </div>
+  );
+};
+
+// Auto-detect current path for 'from' query param so the form can redirect back after cancel
+export const NewEmailButton = ({ copyFrom }: { copyFrom?: string } = {}) => {
+  const params = new URLSearchParams();
+  if (copyFrom) params.set("copy_from", copyFrom);
+  params.set("from", window.location.pathname);
+  const href = `/emails/new?${params.toString()}`;
```
Comment (stefanbinoj) [inline app/javascript/components/EmailsPage/Layout.tsx:106]: Here we set an extra query params for `?from=`. This helps in redirecting back user to correct path from where user clicked action for new or edit email. 

**Why this is necessary?**: 
Consider scenario user is in /published(inertia) and he clicks on new email and redirects to /emails/new (non-inertia) and upon successfull or cancel action user needs to be redirected back to original page from where he came (inertia page)

This jugling from inertia to non-inertia and then to inertia doesn't allow us to have a truthy history thus setting query params from where user came is a simple fix

File: app/javascript/components/EmailsPage/Layout.tsx [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+    <a className={cx("button", { accent: !copyFrom })} href={href}>
+      {copyFrom ? "Duplicate" : "New email"}
+    </a>
+  );
+};
+
+export const EditEmailButton = ({ id }: { id: string }) => {
+  const params = new URLSearchParams();
+  params.set("from", window.location.pathname);
+  const href = `/emails/${id}/edit?${params.toString()}`;
```
Comment (stefanbinoj) [inline app/javascript/components/EmailsPage/Layout.tsx:118]: same reason

File: app/javascript/components/server-components/EmailsPage/EmailForm.tsx [hunk 19]
```diff
@@ -586,33 +586,32 @@ export const EmailForm = () => {
...
+              : installment?.external_id
+                ? "Changes saved!"
+                : "Email created!",
+          "success",
+        );
+        const fromParam = searchParams.get("from");
+        const editUrl = fromParam
+          ? `${editEmailPath(response.installment_id)}?from=${encodeURIComponent(fromParam)}`
+          : editEmailPath(response.installment_id);
+        navigate(editUrl, { replace: true });
```
Comment (stefanbinoj) [inline app/javascript/components/server-components/EmailsPage/EmailForm.tsx:614]: This else block runs for scenarios where the user stays on the edit page:
- "save" - Just saving a draft
- "save_and_preview_email" - Saves and sends preview email
- "save_and_preview_post" - Saves and opens post in new window

```ts
const fromParam = searchParams.get("from");
        const editUrl = fromParam
          ? `${editEmailPath(response.installment_id)}?from=${encodeURIComponent(fromParam)}`
          : editEmailPath(response.installment_id);
        navigate(editUrl, { replace: true });
```
This code block means: When saving, keep the 'from' breadcrumb in the URL so we remember where to go back to when the user clicks Cancel.

**Comment by stefanbinoj** [comment]
Hi @Devwill2, I understand your concern about you working on a related but much larger migration targeting /emails as a whole, while I opened a comparatively smaller PR with a more limited scope (excluding `/new` and `/:id/edit`), which has been approved.

Please consider the following facts:

- All of my PRs take a quite different approach. (first PR introduced an `accumulated client state`, while this PR has a comparatively different architecture and revolves around `infinite scroll`)
- Having open competition helps contributors push each other faster and encourages exploring different approaches.
- My PR hasn’t been merged yet. Even if it does get merged, your PR is still valid and applicable to other scopes as well.

I agree that avoiding duplicated effort is important. In this case, though, the approaches and methods are fundamentally different. My work wasn’t based on or derived from yours, and there was no copying or overlap in implementation.

If the same approach or solution had been duplicated, that would absolutely be a serious issue. But here, the differences in architecture, scope, and strategy mean both efforts explore separate solutions to a similar problem, which can actually be valuable for evaluating trade-offs.

As a final note, I’m happy to comply with whatever the Antiwork team decides.

**Comment by stefanbinoj** [comment]
@EmCousin fixed the failing test 

Screenshot 2025-12-23 at 9 55 42 PM

[fix-1](https://github.com/antiwork/gumroad/pull/2436/files#r2643713640)
[fix-2](https://github.com/antiwork/gumroad/pull/2436/files#r2643821191)

**Comment by EmCousin** [comment]
Thanks for the thoughtful discussion, everyone.

This is an open-source project with a bounty program. Multiple contributors working on the same issue is normal and beneficial; it helps drive better solutions. No one owns an issue until their PR merges, and @stefanbinoj makes a good point about healthy competition.

@Devwill2, your efforts are visible and I understand your frustration. I also went through seven iterations of your work with detailed reviews, which is a significant time investment as well. Where we’re getting stuck is not “small PR vs big PR” as a principle, but that the approach made it hard to review and validate incrementally. The early attempts migrated parts of the page surface without bringing the related endpoints along, and then the scope expanded to cover everything at once. Doing the whole thing end-to-end in one PR is fine if it’s executed cleanly, tested, and self-reviewed, but here we repeatedly ran into the same problems (Inertia patterns not being used as intended, recurring feedback between iterations, and insufficient manual verification).

In contrast, @stefanbinoj’s PR takes a smaller, complete slice and includes the necessary related pieces with it, which makes it easier to test, review, and confidently move forward. It also follows a different implementation approach, so this is not a case of copying someone else’s solution.

And @Devwill2, your work isn’t “wasted”. This PR is not merged yet, and even if it is, there is still remaining scope on the emails migration. Your work can still apply to the remaining parts of the emails page, provided the implementation is complete and all tests pass.

My general advice: the most successful pull requests tend to be scoped as complete, testable units of functionality. Splitting across multiple PRs can also work, as long as the full plan is clear up front and the follow-ups are committed to. The bounty is awarded when the issue is fully resolved.

On the other hand, if a PR goes through multiple rounds and the same issues keep coming back, the risk goes up that someone else will land a working, complete solution first. That’s not a judgment on effort; it’s just that we have limited review bandwidth and the bounty system rewards what ultimately gets merged. If you find yourself in that loop, it’s usually a good signal to pause and reset: narrow the scope, lean on the docs, test thoroughly, and make sure the next iteration fully addresses the feedback before reopening.

I hope this clarifies how we’re approaching this from the reviewer side, and how we can avoid duplicated effort going forward.

**Comment by stefanbinoj** [comment]
@EmCousin _a fixed the failing tests, it was a small fix by adding `ignore_query: true` in url checker as we now rely on `?from=` query params for redirecting user to previous page from where we originally came from. 
Implementation reason discussed [here](https://github.com/antiwork/gumroad/pull/2436#discussion_r2643821191)

Once `/new` and `/:id/edit` pages are also migrated to inertia we can easily redirect user to original page from where he came by using `redirect_back` as inertia would be handling entire history :)

Screenshot 2025-12-24 at 8 14 52 AM

**Comment by stefanbinoj** [comment]
@EmCousin, 
- Failing Test (17) is unrelated 
- Failing Test (11) : 

Screenshot 2025-12-24 at 9 29 18 PM

File: spec/requests/emails/list_spec.rb [hunk 20]
```diff
@@ -221,14 +221,8 @@
         create(:installment, name: "Hello world!", seller:, link: product)
         refresh
 
-        expect(page).to have_table_row({ "Subject" => "Hello world!", "Sent to" => "Customers of Product name", "Audience" => "1" })
-        expect(page).to have_table_row({ "Subject" => "Email 2 (draft)" })
-        expect(page).to_not have_table_row({ "Subject" => "Email 4 (draft)" })
-
-        first("main").scroll_to :bottom
```
Comment (stefanbinoj) [inline spec/requests/emails/list_spec.rb]: This is an edge case of `infinite scroll` 
**Reason** : 
- We already stubbed `PER_PAGE` items `=2`. 
- This means that we can see bottom of table without scrolling thus `infinite scroll` will take place and it and will load the remaining items too. Since there is only 1 more item to load it loads that as well. 

This is a default behaviour of infinite scroll
### ScreenShot

failures_r_spec_example_groups_email_list_emails_draft_emails_paginates_emails_ordered_by_recently_updated_first_with_infinite_scroll_99 2

File: spec/requests/emails/list_spec.rb [hunk 21]
```diff
@@ -117,23 +117,13 @@
...
-        expect(page).to_not have_table_row({ "Subject" => "Hello world!" })
-
-        click_on "Load more"
         wait_for_ajax
-
-        expect(page).to have_table_row({ "Subject" => "Email 1 (sent)" })
-        expect(page).to have_table_row({ "Subject" => "Email 3 (sent)" })
         expect(page).to have_table_row({ "Subject" => "Hello world!" })
-
-        expect(page).to_not have_button("Load more")
```
Comment (stefanbinoj) [inline spec/requests/emails/list_spec.rb:136]: This is a default behaviour of `infinite scroll` 
**Reason** : 
- We already stubbed `PER_PAGE` items `=2`. 
- This means that we can see bottom of table without scrolling thus `infinite scroll` will take place and it and will load the remaining items too. Since there is only 1 more item to load it loads that as well. 

This is a default behaviour of infinite scroll
### ScreenShot

failures_r_spec_example_groups_email_list_emails_draft_emails_paginates_emails_ordered_by_recently_updated_first_with_infinite_scroll_99 2

File: spec/requests/emails/list_spec.rb [hunk 22]
```diff
@@ -117,23 +117,13 @@
...
-
         create(:installment, name: "Hello world!", seller:, link: product, published_at: 10.days.ago)
         refresh
 
         expect(page).to have_table_row({ "Subject" => "Email 1 (sent)" })
         expect(page).to have_table_row({ "Subject" => "Email 3 (sent)" })
-        expect(page).to_not have_table_row({ "Subject" => "Hello world!" })
-
-        click_on "Load more"
         wait_for_ajax
```
Comment (stefanbinoj) [inline spec/requests/emails/list_spec.rb:125]: `infinite scroll` identifies the bottom has been reacched and automatically loads the remaining contents

File: spec/requests/emails/list_spec.rb [hunk 23]
```diff
@@ -279,46 +258,18 @@
...
 
+        # These should not appear
         expect(page).to_not have_table_row({ "Subject" => "Email 2 (draft)" })
         expect(page).to_not have_table_row({ "Subject" => "Email 4 (draft)" })
         expect(page).to_not have_table_row({ "Subject" => "Email 5 (scheduled)" })
         expect(page).to_not have_table_row({ "Subject" => "Email 6 (scheduled)" })
         expect(page).to_not have_table_row({ "Subject" => "Email 7 (sent)" })
         expect(page).to_not have_table_row({ "Subject" => "Hello world" })
-
-        expect(page).to_not have_button("Load more")
```
Comment (stefanbinoj) [inline spec/requests/emails/list_spec.rb:321]: same reason https://github.com/antiwork/gumroad/pull/2436/files#r2647415524

File: spec/requests/emails/list_spec.rb [hunk 24]
```diff
@@ -222,23 +212,13 @@
...
-
-        click_on "Load more"
         wait_for_ajax
-
-        expect(page).to have_table_row({ "Subject" => "Hello world!" })
+        expect(page).to have_table_row({ "Subject" => "Hello world!", "Sent to" => "Customers of Product name", "Audience" => "1" })
         expect(page).to have_table_row({ "Subject" => "Email 2 (draft)" })
         expect(page).to have_table_row({ "Subject" => "Email 4 (draft)" })
-
-        expect(page).to_not have_button("Load more")
```
Comment (stefanbinoj) [inline spec/requests/emails/list_spec.rb:241]: same reason

File: spec/requests/emails/list_spec.rb [hunk 25]
```diff
@@ -279,46 +258,18 @@
...
-        expect(page).to_not have_table_row({ "Subject" => "Email 2 (draft)" })
-        expect(page).to_not have_table_row({ "Subject" => "Email 4 (draft)" })
-        expect(page).to_not have_table_row({ "Subject" => "Email 5 (scheduled)" })
-        expect(page).to_not have_table_row({ "Subject" => "Email 6 (scheduled)" })
-        expect(page).to_not have_table_row({ "Subject" => "Email 7 (sent)" })
-        expect(page).to_not have_table_row({ "Subject" => "Hello world" })
-
-        click_on "Load more"
-        wait_for_ajax
-
```
Comment (Devwill2) [inline spec/requests/emails/list_spec.rb:309]: Hey @stefanbinoj! I have concerns about removing test assertions rather than fixing the infinite scroll behavior.

The Problem
This PR removes important test assertions that verify:

Pagination works correctly
"Load more" button appears/disappears properly
Items load in the correct order
Tests should verify correct behavior, not be modified to pass.

The "Edge Case" Issue
Your comment says infinite scroll auto-loads all items when they fit on screen.

My Understanding
The tests are correct - the implementation needs fixing, not the tests. 
@EmCousin - Am I understanding this correctly? I believe removing tests just to make them pass isn't the right approach here. Would appreciate your thoughts 🙏

Comment (stefanbinoj) [reply spec/requests/emails/list_spec.rb:309]: No you got it wrong, We are not compromising tests here or skipping assertions here! Since we are having `infinite scroll` we no longer need extra checks on `.to have` or `to_not have` on each of actions performed while `click_on "Load more"` as it will have all items automatically fetched automatically. Thus we just removed reduntant checks. We are not removing test assertions here instead updating it for infinite scroll approach.  I hope it clarifies your misunderstanding.

**Comment by stefanbinoj** [comment]
@EmCousin can you trigger a CI for this whenever you are free? Thanks

**Comment by stefanbinoj** [comment]
@EmCousin failing tests seems to be passing locally: 
Screenshot 2025-12-29 at 5 41 16 PM



`emails/list_spec` full testsuite:


Screenshot 2025-12-29 at 5 46 09 PM

**Comment by stefanbinoj** [comment]
@EmCousin all tests seems to be passing!

File: app/javascript/pages/Emails/Drafts.tsx [hunk 26]
```diff
@@ -0,0 +1,169 @@
...
+type AudienceCounts = Map<string, number | "loading" | "failed">;
+
+const audienceCountValue = (audienceCounts: AudienceCounts, installmentId: string) => {
+  const count = audienceCounts.get(installmentId);
+  return count === undefined || count === "loading"
+    ? null
+    : count === "failed"
+      ? "--"
+      : formatStatNumber({ value: count });
+};
```
Comment (EmCousin) [inline app/javascript/pages/Emails/Drafts.tsx:35]: This is duplicate with Scheduled.tsx and could have been factored

**Comment by EmCousin** [comment]
See https://github.com/antiwork/gumroad/pull/2434#issuecomment-3699366472
